From c279503354cf1e2c575d73961415f7a92e67afd6 Mon Sep 17 00:00:00 2001
From: Avinash Kumar <avikumar@habana.ai>
Date: Fri, 12 Feb 2021 11:08:27 +0200
Subject: [PATCH] [SW-28901]: preboot: configure arc as privileged

Required for enabling security by LKD, as the ARC cores
should be able to access the secure registers/memories.

Change-Id: I356a3e3fda490d530a642425d1aacbd54bd120bc
Signed-off-by: Avinash Kumar <avikumar@habana.ai>
---

diff --git a/goya2/Makefile b/goya2/Makefile
index 2158915..4bfd3f5 100644
--- a/goya2/Makefile
+++ b/goya2/Makefile
@@ -1,6 +1,7 @@
 FW_SIZE := 0x4400
 PPBOOT_MAX_SIZE := 18432
 PREBOOT_SRCS_C += $(shell ls preboot/thermal_v2.c)
+PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/security.c)
 
 CFLAGS += -mstrict-align
 
diff --git a/goya2/security.c b/goya2/security.c
new file mode 100644
index 0000000..9644e35
--- /dev/null
+++ b/goya2/security.c
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * (C) Copyright 2020 HabanaLabs, Ltd.
+ * All rights reserved.
+ *
+ */
+#include <preboot.h>
+#include <psoc_arc0_mstr_if_core_hbw_regs.h>
+#include <psoc_arc0_mstr_if_core_lbw_regs.h>
+#include <psoc_arc1_mstr_if_core_hbw_regs.h>
+#include <psoc_arc1_mstr_if_core_lbw_regs.h>
+
+#define EN_ARC_SEC_PRIV		0x3
+
+void preboot_secure_arc(void)
+{
+	/* enable security and privilege mode for arc0 and arc1 */
+	regmap_writel(EN_ARC_SEC_PRIV,
+		mmPSOC_ARC0_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	regmap_writel(EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC0_MSTR_IF_CORE_LBW_AXPROT_OVRD);
+	regmap_writel(EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC1_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	regmap_writel(EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC1_MSTR_IF_CORE_LBW_AXPROT_OVRD);
+}
+
+void preboot_security_init(void)
+{
+	/* TODO add check for security strap ON */
+	preboot_secure_arc();
+}
